keps/sig-cluster-lifecycle/kubeadm/20190925-component-configs.md

```diff
@@ -0,0 +1,223 @@
+- Define strategies for config map naming and backup.
+
+## Proposal
+
+This is where we get down to the nitty gritty of what the proposal actually is.
+
+### Stop Component Config defaulting
+
+Component config defaulting has a number of issues:
+- It’s difficult to tell in retrospect if a setting value was intended or simply defaulted.
+- If a default value changes a defaulted setting will stop us from using the new value.
+- Defaulting bloats the resulting config, thus making it hard for users to migrate it upon a version change.
+
+Hence, kubeadm should stop defaulting component configs and leave this to components themselves.
+
+### Delegate config validation to the components
+
+Currently, kubeadm vendors in pieces of Go code from the components to be able to perform component config validation. This is a bad practice as it bloats the kubeadm dependencies and deviation between the vendored code and the components can lead to inconsistencies.  
+To solve this, validation is going to be performed by executing a container or the binary of the component to be used with an appropriate command to verify the supplied config.  
+If a target component does not have any means to verify its config, then kubeadm won’t do any validation for the time being. It will instead display a warning to the user denoting that certain component’s config is unvalidated.
+
+### Kubernetes Core vs AddOns Component Configs
+
+Currently kubeadm maintains the component configs of both kubelet and kube-proxy. These are components that are maintained in very different ways.
+
+Kubelet, along with API Server, Scheduler, Controller Manager and etcd, are deemed essential Kubernetes components. They are deployed and managed locally by kubeadm, using systemd services (kubelet) or static pods on the local kubelet (API Server, Scheduler, etc.). Hence, their configs are usually consumed by the components in the form of local files.
+
+On the other hand, kube-proxy and CoreDNS/kube-dns are addons, that are managed as standard Kubernetes deployments. Hence their configs are usually accessed via config maps.
+
+Therefore, two different config lifecycles are distinguishable here - one for Kubernetes core components and one for cluster addons.
+
+The defaulting and validation changes, proposed in the previous paragraphs, are going to be applied to both Kubernetes core and AddOn component config types.
+
+#### AddOn Component Config management
+
+AddOn component configs are always stored in config maps. They can be supplied in their entirety by users along with the kubeadm config during init or can be generated by kubeadm itself.
+If, upon upgrade, kubeadm detects an obsolete config version, it will bail out with an error message, asking users to upgrade the addon config by hand.
+
+The ultimate goal for addon component configs, is for them to be eventually managed in their entirety by cluster addon operators, instead of kubeadm.
```

Comment:
some comments.

we added coredns Corefile migration, but it's not an api-machinery component config conversion, per se.
coredns itself does not support conversion (or afaik validation), but kubeadm now embeds the conversion tooling with the idea for this to be an operator at some point.

cc @rajansandeep @chrisohaver 
(for visibility)

until we have an operator for kube-proxy, the kube-proxy CM will remain old and untouched during upgrade.

how would kubeadm detect if this CM is old and unsupported - should it, during upgrade, spawn a container with the new version of kube-proxy and perform validation?

what would be the UX for users to perform the conversion from e.g. v1alpha1 -> 2.
`kube-proxy --convert`, was first rejected during the meting we had about this, which means the users should do it manually, i guess?

Reply:
Kube-proxy's CM will be checked, but not modified during upgrade. If it's too old we can simply refuse the upgrade operation and ask the user to delete the CM (to have it regenerated by kubeadm) or manually upgrade it, and only then re-run `kubeadm upgrade apply`.
Of course, information about this would be available in advance through `kubeadm upgrade plan`.

Perhaps, it's worth adding a non-goal, that this KEP makes no sense for addons, that don't use API machinery for their component configs (such as CoreDNS) and, therefore, are out of scope.